frontend-react/src/hooks/UsePagination.test.ts

```diff
@@ -0,0 +1,210 @@
+import { act, renderHook } from "@testing-library/react-hooks";
+import range from "lodash.range";
+
+import usePagination, {
+    CursorExtractor,
+    getSlotsForResultSet,
+    OVERFLOW_INDICATOR,
+    PaginationActionType,
+} from "./UsePagination";
+
+describe("usePagination", () => {
+    interface SampleRecord {
+        cursor: string;
+    }
+    const cursorExtractor: CursorExtractor<SampleRecord> = (r) => r.cursor;
+
+    function createSampleRecords(numRecords: number): SampleRecord[] {
+        return range(1, numRecords + 1).map(
+            (c) => ({ cursor: c.toString() } as SampleRecord)
+        );
+    }
+
+    test("Hook renders with initial values", () => {
+        const { result } = renderHook(() =>
+            usePagination<SampleRecord>({
+                startCursor: "1",
+                pageSize: 10,
+                cursorExtractor,
+            })
+        );
+
+        expect(result.current.state.fetchStartCursor).toBe("1");
+        // The initial request should check for the presence of up to five pages.
+        expect(result.current.state.fetchCount).toBe(41);
+        expect(result.current.state.currentPageNum).toBe(1);
+        expect(result.current.state.slots).toStrictEqual([]);
+    });
+
+    test("Providing a set of results updates slots", () => {
+        const { result } = renderHook(() =>
+            usePagination<SampleRecord>({
+                startCursor: "1",
+                pageSize: 10,
+                cursorExtractor,
+            })
+        );
+
+        act(() => {
+            result.current.dispatch({
+                type: PaginationActionType.SET_RESULTS,
+                payload: createSampleRecords(10),
+            });
+        });
+        expect(result.current.state.slots).toStrictEqual([1]);
+    });
+
+    test("Setting a page updates the fetch range and num results", () => {
```

Comment:
This test is exemplary of why I looooove having this kind of logic in its own hook!

Reply:
Ha, thanks. I learned a lot about hook testing in this PR.